From a5c1334f1141bbf3dc201d3920c760b850e40eb7 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Thu, 30 May 2024 19:38:57 +0200
Subject: [PATCH 0916/1695] plat-stm32mp1: deprecate stm32_firewall use in scmi
 servers

Deprecate the stm32_firewall use and replace it with the use of
stm32_etzpc_check_ns_access() to check non-secure access to a
peripheral.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 26d7766dcaa677cf0f24550360b17ce3311ea251 plat-stm32mp1: deprecate stm32_firewall use in scmi servers]
Change-Id: Ic6b02493a3c609b6206da50b00991a3c08467749
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/384197
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/scmi_server.c     | 61 ++++++++--------
 .../arm/plat-stm32mp1/scmi_server_scpfw.c     | 69 ++++++++-----------
 2 files changed, 61 insertions(+), 69 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/scmi_server.c b/core/arch/arm/plat-stm32mp1/scmi_server.c
index 93d51c817..145e8e32b 100644
--- a/core/arch/arm/plat-stm32mp1/scmi_server.c
+++ b/core/arch/arm/plat-stm32mp1/scmi_server.c
@@ -13,9 +13,10 @@
 #include <drivers/scmi-msg.h>
 #include <drivers/scmi.h>
 #include <drivers/stm32_cpu_opp.h>
-#include <drivers/stm32_firewall.h>
 #include <drivers/stm32mp_dt_bindings.h>
+#include <drivers/stm32_etzpc.h>
 #include <initcall.h>
+#include <kernel/pm.h>
 #include <mm/core_memprot.h>
 #include <mm/core_mmu.h>
 #include <platform_config.h>
@@ -48,13 +49,13 @@ struct stm32_scmi_clk {
 /*
  * struct stm32_scmi_rd - Data for the exposed reset controller
  * @reset_id: Reset identifier in RCC reset driver
- * @base: Physical controller address
+ * @fw_id: Firewall ID of the peripheral
  * @name: Reset string ID exposed to channel
  * @rstctrl: Reset controller device
  */
 struct stm32_scmi_rd {
 	unsigned long reset_id;
-	paddr_t base;
+	unsigned int fw_id;
 	const char *name;
 	struct rstctrl *rstctrl;
 };
@@ -87,10 +88,10 @@ register_phys_mem(MEM_AREA_IO_NSEC, CFG_STM32MP1_SCMI_SHM_BASE,
 		.enabled = (_init_enabled), \
 	}
 
-#define RESET_CELL(_scmi_id, _id, _base, _name) \
+#define RESET_CELL(_scmi_id, _id, _fw_id, _name) \
 	[(_scmi_id)] = { \
 		.reset_id = (_id), \
-		.base = (_base), \
+		.fw_id = (_fw_id), \
 		.name = (_name), \
 	}
 
@@ -157,26 +158,27 @@ static struct stm32_scmi_clk stm32_scmi_clock[] = {
 
 #ifdef CFG_STM32MP13
 static struct stm32_scmi_rd stm32_scmi_reset_domain[] = {
-	RESET_CELL(RST_SCMI_LTDC, LTDC_R, LTDC_BASE, "ltdc"),
-	RESET_CELL(RST_SCMI_MDMA, MDMA_R, MDMA_BASE, "mdma"),
+	RESET_CELL(RST_SCMI_LTDC, LTDC_R, STM32MP1_ETZPC_LTDC_ID, "ltdc"),
+	RESET_CELL(RST_SCMI_MDMA, MDMA_R, STM32MP1_ETZPC_MAX_ID, "mdma"),
 };
 #endif
 
 #ifdef CFG_STM32MP15
 static struct stm32_scmi_rd stm32_scmi_reset_domain[] = {
-	RESET_CELL(RST_SCMI_SPI6, SPI6_R, SPI6_BASE, "spi6"),
-	RESET_CELL(RST_SCMI_I2C4, I2C4_R, I2C4_BASE, "i2c4"),
-	RESET_CELL(RST_SCMI_I2C6, I2C6_R, I2C6_BASE, "i2c6"),
-	RESET_CELL(RST_SCMI_USART1, USART1_R, USART1_BASE, "usart1"),
-	RESET_CELL(RST_SCMI_STGEN, STGEN_R, STGEN_BASE, "stgen"),
-	RESET_CELL(RST_SCMI_GPIOZ, GPIOZ_R, GPIOZ_BASE, "gpioz"),
-	RESET_CELL(RST_SCMI_CRYP1, CRYP1_R, CRYP1_BASE, "cryp1"),
-	RESET_CELL(RST_SCMI_HASH1, HASH1_R, HASH1_BASE, "hash1"),
-	RESET_CELL(RST_SCMI_RNG1, RNG1_R, RNG1_BASE, "rng1"),
-	RESET_CELL(RST_SCMI_MDMA, MDMA_R, MDMA_BASE, "mdma"),
-	RESET_CELL(RST_SCMI_MCU, MCU_R, 0, "mcu"),
-	RESET_CELL(RST_SCMI_MCU_HOLD_BOOT, MCU_HOLD_BOOT_R, 0,
-		   "mcu_hold_boot"),
+	RESET_CELL(RST_SCMI_SPI6, SPI6_R, STM32MP1_ETZPC_SPI6_ID, "spi6"),
+	RESET_CELL(RST_SCMI_I2C4, I2C4_R, STM32MP1_ETZPC_I2C4_ID, "i2c4"),
+	RESET_CELL(RST_SCMI_I2C6, I2C6_R, STM32MP1_ETZPC_I2C6_ID, "i2c6"),
+	RESET_CELL(RST_SCMI_USART1, USART1_R, STM32MP1_ETZPC_USART1_ID,
+		   "usart1"),
+	RESET_CELL(RST_SCMI_STGEN, STGEN_R, STM32MP1_ETZPC_STGENC_ID, "stgen"),
+	RESET_CELL(RST_SCMI_GPIOZ, GPIOZ_R, STM32MP1_ETZPC_GPIOZ_ID, "gpioz"),
+	RESET_CELL(RST_SCMI_CRYP1, CRYP1_R, STM32MP1_ETZPC_CRYP1_ID, "cryp1"),
+	RESET_CELL(RST_SCMI_HASH1, HASH1_R, STM32MP1_ETZPC_HASH1_ID, "hash1"),
+	RESET_CELL(RST_SCMI_RNG1, RNG1_R, STM32MP1_ETZPC_RNG1_ID, "rng1"),
+	RESET_CELL(RST_SCMI_MDMA, MDMA_R, STM32MP1_ETZPC_MAX_ID, "mdma"),
+	RESET_CELL(RST_SCMI_MCU, MCU_R, STM32MP1_ETZPC_MAX_ID, "mcu"),
+	RESET_CELL(RST_SCMI_MCU_HOLD_BOOT, MCU_HOLD_BOOT_R,
+		   STM32MP1_ETZPC_MAX_ID, "mcu_hold_boot"),
 };
 #endif
 
@@ -555,15 +557,14 @@ int32_t plat_scmi_rd_autonomous(unsigned int channel_id, unsigned int scmi_id,
 				uint32_t state)
 {
 	const struct stm32_scmi_rd *rd = find_rd(channel_id, scmi_id);
-	const struct stm32_firewall_cfg nsec_cfg[] = {
-		{ FWLL_NSEC_RW | FWLL_MASTER(0) },
-		{ }, /* Null terminated */
-	};
+	unsigned int id = 0;
 
 	if (!rd)
 		return SCMI_NOT_FOUND;
 
-	if (rd->base && stm32_firewall_check_access(rd->base, 0, nsec_cfg))
+	id = rd->fw_id;
+
+	if (id != STM32MP1_ETZPC_MAX_ID && stm32_etzpc_check_ns_access(id))
 		return SCMI_DENIED;
 	assert(rd->rstctrl);
 
@@ -591,17 +592,17 @@ int32_t plat_scmi_rd_set_state(unsigned int channel_id, unsigned int scmi_id,
 			       bool assert_not_deassert)
 {
 	const struct stm32_scmi_rd *rd = find_rd(channel_id, scmi_id);
-	const struct stm32_firewall_cfg nsec_cfg[] = {
-		{ FWLL_NSEC_RW | FWLL_MASTER(0) },
-		{ }, /* Null terminated */
-	};
 	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int id = 0;
 
 	if (!rd)
 		return SCMI_NOT_FOUND;
 
-	if (rd->base && stm32_firewall_check_access(rd->base, 0, nsec_cfg))
+	id = rd->fw_id;
+
+	if (id != STM32MP1_ETZPC_MAX_ID && stm32_etzpc_check_ns_access(id))
 		return SCMI_DENIED;
+
 	assert(rd->rstctrl);
 
 	if (assert_not_deassert) {
diff --git a/core/arch/arm/plat-stm32mp1/scmi_server_scpfw.c b/core/arch/arm/plat-stm32mp1/scmi_server_scpfw.c
index 537249a06..dfec6ee02 100644
--- a/core/arch/arm/plat-stm32mp1/scmi_server_scpfw.c
+++ b/core/arch/arm/plat-stm32mp1/scmi_server_scpfw.c
@@ -14,7 +14,7 @@
 #ifdef CFG_STM32_CPU_OPP
 #include <drivers/stm32_cpu_opp.h>
 #endif
-#include <drivers/stm32_firewall.h>
+#include <drivers/stm32_etzpc.h>
 #include <drivers/stm32mp_dt_bindings.h>
 #include <initcall.h>
 #include <mm/core_memprot.h>
@@ -50,14 +50,14 @@ struct stm32_scmi_clk {
 /*
  * struct stm32_scmi_rd - Data for the exposed reset controller
  * @reset_id: Reset identifier in RCC reset driver
- * @base: Physical controller address
+ * @fw_id: Firewall ID of the peripheral
  * @name: Reset string ID exposed to channel
  * @rstctrl: Reset controller manipulated by the SCMI channel
  * @rstctrl_be: Backend reset controller device
  */
 struct stm32_scmi_rd {
 	unsigned long reset_id;
-	paddr_t base;
+	unsigned int fw_id;
 	const char *name;
 	struct rstctrl *rstctrl;
 	struct rstctrl *rstctrl_be;
@@ -86,11 +86,11 @@ struct stm32_scmi_perfd {
 		.change_rate = true, \
 	}
 
-#define RESET_CELL(_scmi_id, _id, _base, _name) \
-	[_scmi_id] = { \
-		.reset_id = _id, \
-		.base = _base, \
-		.name = _name, \
+#define RESET_CELL(_scmi_id, _id, _fw_id, _name) \
+	[(_scmi_id)] = { \
+		.reset_id = (_id), \
+		.fw_id = (_fw_id), \
+		.name = (_name), \
 	}
 
 #define PERFD_CELL(_scmi_id, _name) \
@@ -170,26 +170,27 @@ static struct stm32_scmi_clk stm32_scmi_clock[] = {
 
 #ifdef CFG_STM32MP13
 static struct stm32_scmi_rd stm32_scmi_reset[] = {
-	RESET_CELL(RST_SCMI_LTDC, LTDC_R, LTDC_BASE, "ltdc"),
-	RESET_CELL(RST_SCMI_MDMA, MDMA_R, MDMA_BASE, "mdma"),
+	RESET_CELL(RST_SCMI_LTDC, LTDC_R, STM32MP1_ETZPC_LTDC_ID, "ltdc"),
+	RESET_CELL(RST_SCMI_MDMA, MDMA_R, STM32MP1_ETZPC_MAX_ID, "mdma"),
 };
 #endif
 
 #ifdef CFG_STM32MP15
 static struct stm32_scmi_rd stm32_scmi_reset[] = {
-	RESET_CELL(RST_SCMI_SPI6, SPI6_R, SPI6_BASE, "spi6"),
-	RESET_CELL(RST_SCMI_I2C4, I2C4_R, I2C4_BASE, "i2c4"),
-	RESET_CELL(RST_SCMI_I2C6, I2C6_R, I2C6_BASE, "i2c6"),
-	RESET_CELL(RST_SCMI_USART1, USART1_R, USART1_BASE, "usart1"),
-	RESET_CELL(RST_SCMI_STGEN, STGEN_R, STGEN_BASE, "stgen"),
-	RESET_CELL(RST_SCMI_GPIOZ, GPIOZ_R, GPIOZ_BASE, "gpioz"),
-	RESET_CELL(RST_SCMI_CRYP1, CRYP1_R, CRYP1_BASE, "cryp1"),
-	RESET_CELL(RST_SCMI_HASH1, HASH1_R, HASH1_BASE, "hash1"),
-	RESET_CELL(RST_SCMI_RNG1, RNG1_R, RNG1_BASE, "rng1"),
-	RESET_CELL(RST_SCMI_MDMA, MDMA_R, MDMA_BASE, "mdma"),
-	RESET_CELL(RST_SCMI_MCU, MCU_R, 0, "mcu"),
-	RESET_CELL(RST_SCMI_MCU_HOLD_BOOT, MCU_HOLD_BOOT_R, 0,
-		   "mcu_hold_boot"),
+	RESET_CELL(RST_SCMI_SPI6, SPI6_R, STM32MP1_ETZPC_SPI6_ID, "spi6"),
+	RESET_CELL(RST_SCMI_I2C4, I2C4_R, STM32MP1_ETZPC_I2C4_ID, "i2c4"),
+	RESET_CELL(RST_SCMI_I2C6, I2C6_R, STM32MP1_ETZPC_I2C6_ID, "i2c6"),
+	RESET_CELL(RST_SCMI_USART1, USART1_R, STM32MP1_ETZPC_USART1_ID,
+		   "usart1"),
+	RESET_CELL(RST_SCMI_STGEN, STGEN_R, STM32MP1_ETZPC_STGENC_ID, "stgen"),
+	RESET_CELL(RST_SCMI_GPIOZ, GPIOZ_R, STM32MP1_ETZPC_GPIOZ_ID, "gpioz"),
+	RESET_CELL(RST_SCMI_CRYP1, CRYP1_R, STM32MP1_ETZPC_CRYP1_ID, "cryp1"),
+	RESET_CELL(RST_SCMI_HASH1, HASH1_R, STM32MP1_ETZPC_HASH1_ID, "hash1"),
+	RESET_CELL(RST_SCMI_RNG1, RNG1_R, STM32MP1_ETZPC_RNG1_ID, "rng1"),
+	RESET_CELL(RST_SCMI_MDMA, MDMA_R, STM32MP1_ETZPC_MAX_ID, "mdma"),
+	RESET_CELL(RST_SCMI_MCU, MCU_R, STM32MP1_ETZPC_MAX_ID, "mcu"),
+	RESET_CELL(RST_SCMI_MCU_HOLD_BOOT, MCU_HOLD_BOOT_R,
+		   STM32MP1_ETZPC_MAX_ID, "mcu_hold_boot"),
 };
 #endif
 
@@ -265,18 +266,13 @@ static struct rstctrl plat_resets[ARRAY_SIZE(stm32_scmi_reset)];
 static TEE_Result plat_scmi_reset_assert_level(struct rstctrl *rstctrl,
 					       unsigned int to_us)
 {
-	const struct stm32_firewall_cfg nsec_cfg[] = {
-		{ FWLL_NSEC_PRIV_RW | FWLL_MASTER(0) },
-		{ }, /* Null terminated */
-	};
 	int index = rstctrl - plat_resets;
-	paddr_t domain_iobase = 0;
+	int id = 0;
 
 	assert(index >= 0 && (size_t)index < ARRAY_SIZE(stm32_scmi_reset));
-	domain_iobase = stm32_scmi_reset[index].base;
+	id = stm32_scmi_reset[index].fw_id;
 
-	if (domain_iobase &&
-	    stm32_firewall_check_access(domain_iobase, 0, nsec_cfg))
+	if (id != STM32MP1_ETZPC_MAX_ID && stm32_etzpc_check_ns_access(id))
 		return TEE_ERROR_ACCESS_DENIED;
 
 	return rstctrl_assert_to(stm32_scmi_reset[index].rstctrl_be, to_us);
@@ -285,18 +281,13 @@ static TEE_Result plat_scmi_reset_assert_level(struct rstctrl *rstctrl,
 static TEE_Result plat_scmi_reset_deassert_level(struct rstctrl *rstctrl,
 						 unsigned int to_us)
 {
-	const struct stm32_firewall_cfg nsec_cfg[] = {
-		{ FWLL_NSEC_PRIV_RW | FWLL_MASTER(0) },
-		{ }, /* Null terminated */
-	};
 	int index = rstctrl - plat_resets;
-	paddr_t domain_iobase = 0;
+	int id = 0;
 
 	assert(index >= 0 && (size_t)index < ARRAY_SIZE(stm32_scmi_reset));
-	domain_iobase = stm32_scmi_reset[index].base;
+	id = stm32_scmi_reset[index].fw_id;
 
-	if (domain_iobase &&
-	    stm32_firewall_check_access(domain_iobase, 0, nsec_cfg))
+	if (id != STM32MP1_ETZPC_MAX_ID && stm32_etzpc_check_ns_access(id))
 		return TEE_ERROR_ACCESS_DENIED;
 
 	return rstctrl_deassert_to(stm32_scmi_reset[index].rstctrl_be, to_us);
-- 
2.47.2

